File: app/src/main/java/com/nbennettsoftware/android/npad/PhotoMode.java
package com.nbennettsoftware.android.npad;

import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.Gravity;
import android.view.View;
import android.widget.Toast;

public class PhotoMode {

    private AppCompatActivity activity;
    private View touchOverlay;

    PhotoMode(AppCompatActivity activity){
        this.activity = activity;
    }

    void goIntoPhotoMode(AppCompatActivity activity) {
        View decorView = activity.getWindow().getDecorView();

        decorView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_FULLSCREEN
                | View.SYSTEM_UI_FLAG_IMMERSIVE
                | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION);

        ActionBar actionBar = activity.getSupportActionBar();
        if(actionBar != null) {
            actionBar.hide();
        }

        //editor.hideSoftInput();
        //editor.clearFocus();
        //editor.setFocusable(false);

        touchOverlay.setVisibility(View.VISIBLE);
        touchOverlay.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                exitPhotoMode();
                return true;
            }
        });

        Toast toast = Toast.makeText(activity, R.string.toast_msg_photo_mode_instruction, Toast.LENGTH_SHORT);
        toast.setGravity(Gravity.BOTTOM | Gravity.CENTER, 0, 0);
        toast.show();
    }

    void exitPhotoMode() {
        touchOverlay.setOnClickListener(null);
        touchOverlay.setVisibility(View.GONE);

        //editor.setFocusable(true);
        //editor.setFocusableInTouchMode(true);
        //editor.showSoftInput();

        ActionBar actionBar = activity.getSupportActionBar();
        if(actionBar != null) {
            actionBar.show();
        }

        View decorView = activity.getWindow().getDecorView();
        decorView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_VISIBLE);
    }
}

File: app/src/main/java/com/nbennettsoftware/android/npad/PhotoMode.kt
package com.nbennettsoftware.android.npad

import androidx.appcompat.app.AppCompatActivity
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.Toast

const val HIDE_SYSTEM_WINDOWS_FLAGS = (View.SYSTEM_UI_FLAG_FULLSCREEN
        or View.SYSTEM_UI_FLAG_IMMERSIVE
        or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION)

const val EXIT_TOUCH_FRAME_TAG = "photoMode_exitTouchFrame"

class PhotoMode (private val activity: AppCompatActivity,
                 private val rootView: ViewGroup,
                 private val contentView : ViewGroup,
                 private val controlsView: View,
                 private val hideSystemWindows : Boolean) {

    private val view = rootView//A random view reference to access context etc...
    private var touchFrame: FrameLayout? = null
    private var systemUiFlagsBackup = view.systemUiVisibility
    private var systemWindowsHidden = false

    fun enterPhotoMode() {
        hideControls()
        hideKeyboard()
        defocusEditText()
        setContentViewUnfocusable()
        if(hideSystemWindows) doHideSystemWindows()
        applyPhotoModeExitOverlay()
    }

    private fun hideKeyboard(){
        val imm = activity.getSystemService(AppCompatActivity.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(view.windowToken, 0)
    }

    private fun hideControls() {
        controlsView.visibility = View.INVISIBLE
    }

    private fun showControls() {
        controlsView.visibility = View.VISIBLE
    }

    private fun defocusEditText() {
        val focusedView = view.findFocus()
        if(focusedView is EditText) {
            focusedView.clearFocus()
        }
    }

    private fun doHideSystemWindows(){
        systemUiFlagsBackup = view.systemUiVisibility
        view.systemUiVisibility = HIDE_SYSTEM_WINDOWS_FLAGS
        systemWindowsHidden = true
    }

    private fun unhideSystemWindows(){
        view.systemUiVisibility = systemUiFlagsBackup
    }

    private fun setContentViewUnfocusable(){
        contentView.isFocusable = false
        contentView.isFocusableInTouchMode = false
    }

    private fun setContentViewFocusable() {
        contentView.isFocusable = true
        contentView.isFocusableInTouchMode = false
    }

    private fun applyPhotoModeExitOverlay() {
        touchFrame = FrameLayout(view.context)
        touchFrame!!.tag = EXIT_TOUCH_FRAME_TAG
        val layoutParam = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT)

        touchFrame!!.setOnLongClickListener{
            exitPhotoMode()
            true
        }

        rootView.addView(touchFrame, layoutParam)

        val toast = Toast.makeText(activity, R.string.toast_msg_photo_mode_instruction, Toast.LENGTH_SHORT)
        toast.setGravity(Gravity.BOTTOM or Gravity.CENTER, 0, 0)
        toast.show()
    }

    private fun removePotentialPhotoModeExitOverlay() {
        if(touchFrame == null) return
        rootView.removeView(touchFrame)
        touchFrame = null
    }

    private fun exitPhotoMode() {
        showControls()
        setContentViewFocusable()
        if(systemWindowsHidden) unhideSystemWindows()
        removePotentialPhotoModeExitOverlay()
    }
}
